File: two-sum/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        
        # creating dictionary with complements:
        complements = {num: ix for ix, num in enumerate(nums)}
        
        # if complemented number is found in dictionary and is different, return pair:
        for ix, num in enumerate(nums):
            if (target - num) in complements.keys() and ix != complements[target - num]:
                return [ix, complements[target - num]]

File: two-sum/two-sum.cpp
class Solution {
public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        
        // creating a set of complements:
        std::map<int, int> comp;
        for (int ix=0; ix<nums.size(); ix++) {
            comp.insert(std::pair<int, int>(nums[ix], ix));
        }
        
        // iterate over all elements and find if there exists a complement:
        for (int ix=0; ix<nums.size(); ix++) {
            if (comp.find(target - nums[ix]) != comp.end()) {
                if (ix != comp[target - nums[ix]]) {
                    return std::vector<int>({ix, comp[target - nums[ix]]});
                }
            }
        }
        
        return std::vector<int>({-1, -1});
    }
};
